src/dragonball/src/vcpu/vcpu_manager.rs

```diff
@@ -0,0 +1,1005 @@
+    tid: u32,
+}
+
+impl std::fmt::Debug for VcpuInfo {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        f.debug_struct("VcpuInfo")
+            .field("vcpu", &self.vcpu.is_some())
+            .field("vcpu_fd", &self.vcpu_fd.is_some())
+            .field("handle", &self.handle.is_some())
+            .field("tid", &self.tid)
+            .finish()
+    }
+}
+
+/// Manage all vcpu related actions
+pub struct VcpuManager {
+    pub(crate) vcpu_infos: Vec<VcpuInfo>,
+    vcpu_config: VcpuConfig,
+    vcpu_seccomp_filter: BpfProgram,
+    vcpu_state_event: EventFd,
+    vcpu_state_sender: Sender<VcpuStateEvent>,
+    support_immediate_exit: bool,
+
+    // The purpose of putting a reference of IoManager here is to simplify the
+    // design of the API when creating vcpus, and the IoManager has numerous OS
+    // resources that need to be released when vmm exits. However, since
+    // VcpuManager is referenced by VcpuEpollHandler and VcpuEpollHandler will
+    // not be released when vmm is closed, we need to release io manager
+    // manually when we exit all vcpus.
+    io_manager: Option<IoManagerCached>,
+    shared_info: Arc<RwLock<InstanceInfo>>,
+    vm_as: GuestAddressSpaceImpl,
+    pub(crate) vm_fd: Arc<VmFd>,
+
+    action_sycn_tx: Option<Sender<bool>>,
+    vcpus_in_action: (VcpuAction, Vec<u8>),
+    pub(crate) reset_event_fd: Option<EventFd>,
+
+    #[cfg(feature = "hotplug")]
+    upcall_channel: Option<Arc<UpcallClient<DevMgrService>>>,
+
+    // X86 specific fields.
+    #[cfg(target_arch = "x86_64")]
+    pub(crate) supported_cpuid: kvm_bindings::CpuId,
+}
+
+#[allow(clippy::too_many_arguments)]
+impl VcpuManager {
+    /// Get a new VcpuManager instance
+    pub fn new(
+        vm_fd: Arc<VmFd>,
+        kvm_context: &KvmContext,
+        vm_config_info: &VmConfigInfo,
+        vm_as: GuestAddressSpaceImpl,
+        vcpu_seccomp_filter: BpfProgram,
+        shared_info: Arc<RwLock<InstanceInfo>>,
+        io_manager: IoManagerCached,
+        epoll_manager: EpollManager,
+    ) -> Result<Arc<Mutex<Self>>> {
+        let support_immediate_exit = kvm_context.kvm().check_extension(Cap::ImmediateExit);
+        let max_vcpu_count = vm_config_info.max_vcpu_count;
+        let kvm_max_vcpu_count = kvm_context.get_max_vcpus();
+
+        // check the max vcpu count in kvm. max_vcpu_count is u8 and kvm_context.get_max_vcpus()
+        // returns usize, so convert max_vcpu_count to usize instead of converting kvm max vcpu to
+        // u8, to avoid wraping usize. Otherwise if kvm_max_vcpu_count is greater than 255, it'll
+        // be casted into a smaller number.
+        if max_vcpu_count as usize > kvm_max_vcpu_count {
+            error!(
+                "vcpu_manager: specified vcpu count {} is greater than max allowed count {} by kvm",
+                max_vcpu_count, kvm_max_vcpu_count
+            );
+            return Err(VcpuManagerError::MaxVcpuLimitation);
```

Comment:
The error message is "richer" than the error here. You could add the 2 counts to the `MaxVcpuLimitation` and return that. But since the errortype suports debug, you could also log the actual error with `error!()` potentially.

Reply:
Thanks, I've added the 2 counts into the `MaxVcpuLimitation` error:)